neps/nep-0509.md

```diff
@@ -0,0 +1,592 @@
+
+So, the actual reward never exceeds total reward, and when everyone does perfect work, they are equal.
+For the context of the NEP, it is enough to assume that `work_quality_ratio = avg_{role}({role}_quality_ratio)`.
+So, if node is both a block and chunk producer, we compute quality for each role separately and then take average of them.
+  
+When epoch is finalized, all block headers in it uniquely determine who was expected to produce each block and chunk.
+Thus, if we define quality ratio for block producer as `produced_blocks/expected_blocks`, everyone is able to compute it.
+Similarly, `produced_chunks/expected_chunks` is a quality for chunk producer. 
+It is more accurate to say `included_chunks/expected_chunks`, because inclusion of chunk in block is a final decision of a block producer which defines success here.
+
+Ideally, we could compute quality for chunk validator as `produced_endorsements/expected_endorsements`. Unfortunately, we won't do it in Stage 0 because:
+
+* Mask of endorsements is not part of the block header, and it would be a significant change;
+* Block producer doesn't have to wait for all endorsements to be collected, so it could be unfair to say that endorsement was not produced if block producer just went ahead.
+
+So for now we decided to compute quality for chunk validator as ratio of `included_chunks/expected_chunks`, where we iterate over chunks which node was expected to validate.
+The obvious drawback here is that if chunks are not produced at all, chunk validators will also be impacted. We plan to address it in the future releases.
+ 
+#### Kickouts
+
+In addition to that, if node performance is too poor, we want a mechanism to kick it out of the validator list, to ensure healthy protocol performance and validator rotation.
+Currently, we have a threshold for each role, and if for some role the same `{role}_quality_ratio` is lower than threshold, the node is kicked out.
+
+If we write this in pseudocode,
+
+```python
+if validator is block producer and block_producer_quality_ratio < 0.8:
+    kick out validator
+if validator is chunk producer and chunk_producer_quality_ratio < 0.8:
+    kick out validator
+```
+
+For chunk validator, we apply absolutely the same formula. However, because:
+
+* the formula doesn't count endorsements explicitly 
+* for chunk producers it kind of just makes chunk production condition stronger without adding value
+
+we apply it to nodes which **only validate chunks**. So, we add this line:
+
+```python
+if validator is only chunk validator and chunk_validator_quality_ratio < 0.8:
+    kick out validator
+```
+
+As we pointed out above, current formula `chunk_validator_quality_ratio` is problematic.
+Here it brings even a bigger issue: if chunk producers don't produce chunks, chunk validators will be kicked out as well, which impacts network stability.
+This is another reason to come up with the better formula.  
```

Comment:
Chunk validators can collude and not endorse some chunks in a way that some chunk producers or other chunk validators get kicked out by not getting their chunks included.

Reply:
But this is more relevant to the chunk endorsement process, not chunk validator kickouts/rewards.

And the base assumption of new approach is to make event "1/3 validators of chunk collude" mean that "1/3 of all validators collude" with high probability, so in this case the base blockchain security assumption fails, on which we rely on.